### PR TITLE
Query: Optimize OfType operator to only query for required columns

### LINKED ISSUE #7451: Relational OfType op selects unnecessary columns
Given a hierarchy like:
```
   A
 /   \
B     C
```
the following query will select the columns from B:
```c#
context.As.OfType<C>();
```

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #7451 

For OfType operators of entity type remove the operator by changing the type of entityqueryable.
Move TypeIs to query model level optimization

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 1]
```diff
@@ -397,22 +397,34 @@ public override void VisitQueryModel(QueryModel queryModel)
...
             }
+        }
 
-            private class RecursiveQueryModelExpressionVisitor : ExpressionVisitorBase
-            {
-                private readonly NondeterministicResultCheckingVisitor _parentVisitor;
+        /// <summary>
+        /// 
+        /// </summary>
+        protected class RecursiveQueryModelExpressionVisitor : ExpressionVisitorBase
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: @anpete - Any place to put this visitor? It doesn't need to be on public surface. (May somewhere in .Internal)

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 2]
```diff
@@ -1306,6 +1317,80 @@ protected virtual void WarnClientEval([NotNull] object expression)
                 () => RelationalStrings.ClientEvalWarning(expression));
         }
 
+        private class OfTypeResultOperatorOptimizer : QueryModelVisitorBase
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: Any reason not to do this in Core? If not, then QueryOptimizer would be the natural place to put this logic.

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: I guess this can be done in core too. I started in relational because extra columns/select expression is in relational. 😄

**Comment by smitpatel** [comment]
New version up
- Change entityqueryable to the desired type to avoid selecting columns from sibling types
- Remove OfType of entity type is applied eagerly and removed from query model
- Lift predicate from join table when Inner Join to produce flattened query
- Convert join table to subquery in the presence of predicate in Left Outer Join
- Move TypeIsExpressionVisitor to relational OptimizeQueryModel method

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs [hunk 3]
```diff
@@ -637,111 +635,8 @@ private static Expression HandleMax(HandlerContext handlerContext)
...
-                    && discriminatorExpression.QuerySource == _querySource)
-                {
-                    return new DiscriminatorPredicateExpression(_discriminatorPredicate, _querySource);
-                }
-
-                return expression;
-            }
-        }
+        // OfType of an entity type should be removed while optimizing query model
+        private static Expression HandleOfType(HandlerContext handlerContext) => handlerContext.EvalOnClient();
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: Can remove this and the corresponding entry in the handler map because client-eval is the default.

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 4]
```diff
@@ -397,22 +399,97 @@ public override void VisitQueryModel(QueryModel queryModel)
...
+            => Expression.Constant(
+                _createEntityQueryableMethod
+                    .MakeGenericMethod(targetEntityType.ClrType)
+                    .Invoke(null, new object[]
+                    {
+                        entityQueryProvider
+                    }));
+
+            private static readonly MethodInfo _createEntityQueryableMethod
+                = typeof(NavigationRewritingExpressionVisitor)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: Wrong type?

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 5]
```diff
@@ -397,22 +399,97 @@ public override void VisitQueryModel(QueryModel queryModel)
...
+                base.VisitQueryModel(queryModel);
 
-                public RecursiveQueryModelExpressionVisitor(NondeterministicResultCheckingVisitor parentVisitor)
-                {
-                    _parentVisitor = parentVisitor;
-                }
+                var ofTypeOperators = queryModel.ResultOperators.OfType<OfTypeResultOperator>().ToList();
 
-                protected override Expression VisitSubQuery(SubQueryExpression expression)
+                foreach (var resultOperator in ofTypeOperators)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: Can this be done in VisitResultOperator in QueryOptimizer? If so, we can DRY up some of the code below that deals with updating QSREs and annotations.

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 6]
```diff
@@ -397,22 +399,97 @@ public override void VisitQueryModel(QueryModel queryModel)
...
+                            queryAnnotation.QuerySource = newMainFromClause;
+                            queryAnnotation.QueryModel = queryModel;
+                        }
+                    }
                 }
+
+                queryModel.TransformExpressions(new RecursiveQueryModelExpressionVisitor(this).Visit);
+            }
+
+            private ConstantExpression CreateEntityQueryable(IEntityType targetEntityType, IAsyncQueryProvider entityQueryProvider)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: DRY this up by moving this and the nav rewrite version to a shared helper. Perhaps IAsyncQueryProviderExtensions.CreateQueryable?

File: src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs [hunk 7]
```diff
@@ -397,22 +399,97 @@ public override void VisitQueryModel(QueryModel queryModel)
...
+                = typeof(NavigationRewritingExpressionVisitor)
+                    .GetTypeInfo().GetDeclaredMethod(nameof(_CreateEntityQueryable));
+
+            [UsedImplicitly]
+            // ReSharper disable once InconsistentNaming
+            private static EntityQueryable<TResult> _CreateEntityQueryable<TResult>(IAsyncQueryProvider entityQueryProvider)
+                => new EntityQueryable<TResult>(entityQueryProvider);
+        }
+
+        private class RecursiveQueryModelExpressionVisitor : ExpressionVisitorBase
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/EntityQueryModelVisitor.cs]: If we add this, we should use it everywhere. Can be a separate PR.

**Comment by smitpatel** [comment]
Updated.

File: src/Microsoft.EntityFrameworkCore/Query/Internal/QueryOptimizer.cs [hunk 8]
```diff
@@ -248,6 +253,50 @@ public override void VisitResultOperator(ResultOperatorBase resultOperator, Quer
...
+                        {
+                            queryModel.ResultOperators.RemoveAt(index);
+
+                            var newMainFromClause = new MainFromClause(
+                                oldQuerySource.ItemName,
+                                entityType.ClrType,
+                                entityQueryProvider.CreateEntityQueryable(entityType));
+                            queryModel.MainFromClause = newMainFromClause;
+
+                            var querySourceMapping = new QuerySourceMapping();
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/QueryOptimizer.cs]: Can you unify any of this fix-up stuff with similar code in this class?

File: src/Microsoft.EntityFrameworkCore/Query/Internal/IQueryOptimizer.cs [hunk 9]
```diff
@@ -20,6 +21,7 @@ public interface IQueryOptimizer
         /// </summary>
         void Optimize(
             [NotNull] IReadOnlyCollection<IQueryAnnotation> queryAnnotations,
+            [NotNull] IModel model,
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/IQueryOptimizer.cs]: Make this a ctor param.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 10]
```diff
@@ -1294,6 +1298,20 @@ public override void VisitResultOperator(ResultOperatorBase resultOperator, Quer
         }
 
         /// <summary>
+        ///     Applies optimizations to the query.
+        /// </summary>
+        /// <param name="queryModel"> The query. </param>
+        protected override void OptimizeQueryModel(QueryModel queryModel)
+        {
+            var typeIsExpressionTranslatingVisitor
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: Does this code need to move? This isn't really an optimization but rather a specific translation.

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: It causes indirect optimization in terms of materialization. We have a test with `a is Kiwi` in projection. If this is done in optimization we don't materialize. If we do during VisitQueryModel, we materialize the entity.
